### DISCUSSION TITLE
List DAG Run and List Task Instance empty with DAG-level access control

### DISCUSSION THREAD

**Discussion by kmehkeri** [apache/airflow#21858] [General] [answered]
### Apache Airflow version

2.2.2

### What happened

User has role with following permissions:

> can read on Connections, can read on DAG Runs, can read on Task Instances, can read on Audit Logs, can read on ImportError, can read on Pools, can read on Variables, can read on XComs, can read on DAG Code, can read on DAG Dependencies, can read on My Password, can edit on My Password, can read on My Profile, can edit on My Profile, can read on SLA Misses, can read on Task Logs, can read on Website, can read on Task Reschedules, can read on DAG:hello

It can only see DAG named `hello` in the DAG List. It can get tree, graph, details, logs, etc. of that DAG only, but not the other DAGs. It can retrieve DAG Runs and Task Instances for this specific DAG via API - all as expected.

However, List DAG Run and List Task Instance views display no results, both when navigating from menu and from the links in Runs / Recent Tasks columns on the DAG list.

### What you expected to happen

List DAG Run and List Task Instance views should display DAG runs and task instances (limited to the DAGs user has access to).

### How to reproduce

1. Create a role with permissions as specified in "What happened" (can read on everything apart from can read on DAG).
2. Create a DAG: 

```
def hello():
    print("Hello world!")

with DAG('hello', description='Hello World', start_date=datetime(2022, 1, 1), access_control={'my_role': ['can_read']}) as dag:
    hello = PythonOperator(task_id='hello', python_callable=hello)

```
3. Navigate to Browse->List DAG Run and Browse->List Task Instance

### Operating System

That's Airflow deployed from official Helm chart on AKS

### Deployment

Official Apache Airflow Helm Chart

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by kmehkeri** [comment]
I noticed that both lists are displayed correctly with edit permission on the DAG, ie.: `access_control={'my_role': ['can_read', 'can_edit']}`.

**Comment by potiuk** [comment]
I am not sure if this is wrong, so I converted it into a discussion. But if you want to track it down and provide a PR for that - feel free.

**Comment by kmehkeri** [comment]
I'm happy to provide a PR, but so far I was unable to track where the trouble is. Here's what I have:

In `airflow/www/views.py` we have [DagRunModelView](https://github.com/apache/airflow/blob/f25a58eebeb9ad3283fca2daa6666811f1a036c6/airflow/www/views.py#L3934) and [TaskInstanceModelView](https://github.com/apache/airflow/blob/f25a58eebeb9ad3283fca2daa6666811f1a036c6/airflow/www/views.py#L4249) with:
```
base_filters = [['dag_id', DagFilter, lambda: []]]
```
where [DagFilter](https://github.com/apache/airflow/blob/f25a58eebeb9ad3283fca2daa6666811f1a036c6/airflow/www/views.py#L3177) has:
```
filter_dag_ids = current_app.appbuilder.sm.get_accessible_dag_ids(g.user)
```
The `get_accessible_dag_ids` should pick all DAGs a user has read access to, same is used in [index](https://github.com/apache/airflow/blob/f25a58eebeb9ad3283fca2daa6666811f1a036c6/airflow/www/views.py#L613) to display DAG list.

There's also [AirflowPrivilegeVerifierModelView](https://github.com/apache/airflow/blob/f25a58eebeb9ad3283fca2daa6666811f1a036c6/airflow/www/views.py#L3196), which verifies edit access, but if I read correctly this should only apply to actions/operations, not listing.

Honestly this is where I'm stuck.

And why did you actually mention that you're not sure if it's wrong?

**Reply by potiuk** [reply]
@jhtimmins - maybe you can take a look ? I think that's something you are intimately familiar with :)

**Reply by potiuk** [reply]
> And why did you actually mention that you're not sure if it's wrong?

Becasue I am not :). I do not know all details and all the assumptions for the permission model. Maybe there is a good reason for this.

**Comment by kmehkeri** [comment, marked as answer]
Okay, I finally dug it out. This was a bug introduced by #16634 in 2.2.0 and fixed by #20663 in 2.2.4. I was using 2.2.2, upgrading the image fixes it - case closed.
